src/stan-users-guide/parallelization.Rmd

```diff
@@ -1,12 +1,314 @@
+requires setting a sensible tuning parameter for good performance.
+
+The tuning parameter is the so-called `grainsize`. For the
+`reduce_sum` version the `grainsize` is merely a suggested partial sum
+size while for the `reduce_sum_static` version the `grainsize`
+specifies the maximal partial sum size. While for `reduce_sum` a
+`grainsize` of one commonly leads to good performance already (since
+automatic aggregation is performed), the `reduce_sum_static` variant
+requires setting a sensible `grainsize` for good performance as
+explained in [more detail below](#reduce-sum-grainsize).
+
+For efficiency and convenience additional
+shared arguments can be passed to every term in the sum. So for the
+array ```{ x1, x2, ... }``` and the shared arguments ```s1, s2, ...```
+the effective sum (with individual terms) looks like: 
+
+```
+g(x1, s1, s2, ...) + g(x2, s1, s2, ...) + g(x3, s1, s2, ...) + ...
+```
+
+which can be written equivalently with partial sums to look like:
+
+```
+f({ x1, x2 }, s1, s2, ...) + f({ x3 }, s1, s2, ...)
+```
+
+where the particular slicing of the ```x``` array can change.
+
+Given this, the signatures are:
+
+```
+real reduce_sum(F f, T[] x, int grainsize, T1 s1, T2 s2, ...)
+real reduce_sum_static(F f, T[] x, int grainsize, T1 s1, T2 s2, ...)
+```
+
+1. ```f``` - User defined function that computes partial sums
+2. ```x``` - Array to slice, each element corresponds to a term in the summation
+3. ```grainsize``` - Target for size of slices
+4. ```s1, s2, ...``` - Arguments shared in every term
+
+The user-defined partial sum functions have the signature:
+
+```
+real f(int start, int end, T[] x_slice, T1 s1, T2 s2, ...)
+```
+
+and take the arguments:
+
+1. ```start``` - An integer specifying the first term in the partial sum
+2. ```end``` - An integer specifying the last term in the partial sum (inclusive)
+3. ```x_slice``` - The subset of ```x``` (from ```reduce_sum``` / `reduce_sum_static`) for
+which this partial sum is responsible (```x_slice = x[start:end]```)
+4. ```s1, s2, ...``` - Arguments shared in every term  (passed on
+without modification from the ```reduce_sum``` / `reduce_sum_static` call)
+
+The user-provided function ```f``` is expected to compute the partial
+sum with the terms ```start``` through ```end``` of the overall
+sum. The user function is passed the subset ```x[start:end]``` as
+```x_slice```. ```start``` and  ```end``` are passed so that ```f```
+can index any of the tailing ```sM``` arguments as necessary. The
+trailing ```sM``` arguments are passed without modification to every
+call of ```f```.
+
+A ```reduce_sum``` (or `reduce_sum_static`) call:
+
+```
+real sum = reduce_sum(f, x, grainsize, s1, s2, ...);
+```
+
+can be replaced by either:
+
+```
+real sum = f(1, size(x), x, s1, s2, ...);
+```
+
+or the code:
+
+```
+real sum = 0.0;
+for(i in 1:size(x)) {
+  sum += f(i, i, { x[i] }, s1, s2, ...);
+}
+```
+
+### Example: Logistic Regression
```

Comment:
in this section, every paragraph is a single sentence - I don't think this is needed.